hopr-socks/src/main.rs

```diff
@@ -0,0 +1,203 @@
+    };
+}
+
+async fn spawn_socks_client(
+    socks_domain: &String,
+    target_host: String,
+    target_port: u16,
+    auth: AuthMode,
+) -> Result<()> {
+    let remote_domain: String = [target_host.clone(), target_port.to_string()].join(":");
+    let config = ClientConfig::default();
+
+    // Creating a SOCKS stream to the target address through the socks server
+    let mut socks = match auth {
+        AuthMode::NoAuth => Socks5Stream::connect(socks_domain.clone(), target_host, target_port, config).await?,
+        AuthMode::Password { username, password } => {
+            Socks5Stream::connect_with_password(
+                socks_domain.clone(),
+                target_host,
+                target_port,
+                username,
+                password,
+                config,
+            )
+            .await?
+        }
+    };
+
+    // Once connection is completed, can start to communicate with the server
+    http_request(&mut socks, remote_domain).await?;
+
+    Ok(())
+}
+async fn spawn_socks_server(socks_domain: &String, timeout: u64, auth: AuthMode) -> Result<()> {
+    let mut config = ServerConfig::default();
+    config.set_request_timeout(timeout);
+
+    let config = match auth {
+        AuthMode::NoAuth => {
+            warn!("No authentication has been set!");
+            config
+        }
+        AuthMode::Password { username, password } => {
+            info!("Simple auth system has been set.");
+            config.with_authentication(SimpleUserPassword { username, password })
+        }
+    };
+
+    let config = Arc::new(config);
+    let listener = TcpListener::bind(&socks_domain).await?;
+
+    info!("Listen for socks connections @ {}", &socks_domain);
+
+    // Standard TCP loop
+    loop {
+        match listener.accept().await {
+            Ok((socket, _addr)) => {
+                info!("Connection from {}", socket.peer_addr()?);
+                let socket = Socks5Socket::new(socket, config.clone());
+
+                spawn_and_log_error(socket.upgrade_to_socks5());
+            }
+            Err(err) => error!("accept error = {:?}", err),
+        }
+    }
+}
+
+fn spawn_and_log_error<F, T, A>(fut: F) -> task::JoinHandle<()>
+where
+    F: Future<Output = Result<Socks5Socket<T, A>>> + Send + 'static,
+    T: AsyncRead + AsyncWrite + Unpin,
+    A: Authentication,
+{
+    task::spawn(async move {
+        if let Err(e) = fut.await {
+            error!("{:#}", &e);
+        }
+    })
+}
+
+async fn http_request<T: AsyncRead + AsyncWrite + Unpin>(stream: &mut T, domain: String) -> Result<()> {
+    debug!("Requesting body...");
+
+    // construct our request, with a dynamic domain
+    let mut headers = vec![];
+    headers.extend_from_slice("GET / HTTP/1.1\n".as_bytes());
+    headers.extend_from_slice(format!("Host: {domain}\n").as_bytes());
+    headers.extend_from_slice("User-Agent: hopr-socks/0.1.0\n".as_bytes());
+    headers.extend_from_slice("Accept: */*\n\n".as_bytes());
```

Comment:
This should go to the test though, it should not be in main.

Reply:
Related to comment above.

Reply:
Removed.